docs/linking.md

```diff
@@ -0,0 +1,218 @@
+
+```
+-ir-compression lite
+-ir-compression none
+```
+
+They are also available via the API through the `spProcessCommandLineArguments` function. 
+
+Note that Slang can consume and process either `lite` or `none` styles transparently. Also mixing compressed libraries with uncompressed libraries also works. 
+
+Symbols
+-------
+
+Linkage is the process of combining of libraries and source to produce something. The mechanism used with Slang, is that items that are going to be imported or exported are named. As the Slang language supports function overloading and generics, it is not possible to just use a function name, or type name to uniquely identify the item. Thus exported and imported names are 'mangled' to include this additional information. This is not unlike how C++ produces mangled names for symbols for similar reasons.
+
+Symbols within a module are additionally uniquely marked with the name of the module, and thus the module name becomes part of the mangled name. This has implications for includes. For example say we have two modules, and they both #include a header file that contains a function. That function will appear as two separate functions - one defined in each module, because a #include is a textual the function will be defined once in each module and with each module name.
+
+By default Slang will provide linkage symbols for all global symbols - types or functions, including declarations. In the future this is likely to change with declarations *not* being exported by default, as the declaration is made available just to forward define within the current compilation, not as a mechnism to descibe something to be imported. 
+
+To mark a declaration as having it's definition defined elsewhere outside the current module, the `[__extern]` attribute can be used. For example 
+
+```
+[__extern] Thing makeThing(int a, int b);
+``` 
+
+With types the `[__extern]` attribute can also be used. For the moment though a type declaration that is externed still needs to be a type definition. Therefore to declare a type to be imported from elsewhere we actually have to write
+
+```
+[__extern] struct Thing {};
+```
+
+Which looks a little like a definition of `Thing` but because of the `[__extern]` it actually means that `Thing` is defined elsewhere.  
+
+Without further language and API support, having every module have all of it's symbols include the module name becomes a problem. If we have two modules `module-a` and `module-b` - how do I use a function defined in `module-a` in `module-b`? There needs to be mechansism/s to allow access to the function as part of the other module. Until the language and API features are added to allow such support, the problem is side stepped by allowing the specifying of a module name. If `module-a` and `module-b` use the same module name they will be able to access one anothers symbols during linkage. 
```

Comment:
Hmm... I'm not sure this verbiage is 100% correct. The case where you have different module names is actually the one that should Just Work. If I have:

```
// A.slang
float foo(float x) { return x; }
```

and:

```
// B.slang
import A;

float bar(float x) { return foo(x); }
```

Then I should be able to compile `A.slang` to produce `A.slang-lib` and `B.slang` to produce `B.slang-lib`, and then link them, and it should Just Work without any `[__export]s` or need to fiddle with module names. This is how things are designed to work.

The reason this doesn't work for some of the scenarios we are trying to support today are:

1. The client code isn't using `import` right now, so everything has to find a way to work with the `#include` model and a single module name.

2. Even if we switched the code to use `import` then because we are trying to plug in multiple variants of module `A` (say), we end up needing multiple definitions of `A`: a "stub" one to use when compiling `B` offline, and then multiple "real" ones to link it against.

Neither of these is really a thing we want to support in the long run, and I wouldn't want the documentation here to overly prioritize describing the workaround case when the "clean" case is actually the one that is closest to working as intended today.